osquery/carver/carver.cpp

```diff
@@ -0,0 +1,286 @@
+  if (!ret) {
+    LOG(ERROR) << "Unable to create carve file store";
+  }
+
+  // Store the path to our archive for later exfiltration
+  archivePath_ = carveDir_ / fs::path(kCarveNamePrefix + carveGuid_ + ".tgz");
+
+  // Update the DB to reflect that the carve is pending.
+  updateCarveValue(carveGuid_, "status", "PENDING");
+};
+
+Carver::~Carver() {
+  fs::remove_all(carveDir_);
+}
+
+void Carver::start() {
+  for (const auto& p : carvePaths_) {
+    if (!fs::exists(p)) {
+      LOG(WARNING) << "File does not exist on disk: " << p;
+    } else {
+      Status s = carve(p);
+      if (!s.ok()) {
+        LOG(WARNING) << "Failed to carve file: " << p;
+      }
+    }
+  }
+
+  std::set<fs::path> carvedFiles;
+  for (const auto& p : platformGlob((carveDir_ / "*").string())) {
+    carvedFiles.insert(fs::path(p));
+  }
+
+  auto s = compress(carvedFiles);
+  if (!s.ok()) {
+    LOG(WARNING) << "Failed to create carve archive: " << s.getMessage();
+    updateCarveValue(carveGuid_, "status", "FAILED");
+    return;
+  }
+
+  s = exfil(archivePath_);
+  if (!s.ok()) {
+    LOG(WARNING) << "Failed to post carve: " << s.getMessage();
+    updateCarveValue(carveGuid_, "status", "FAILED");
+    return;
+  }
+};
+
+Status Carver::carve(const boost::filesystem::path& path) {
+  PlatformFile src(path.string(), PF_OPEN_EXISTING | PF_READ);
+  PlatformFile dst((carveDir_ / path.leaf()).string(),
+                   PF_CREATE_NEW | PF_WRITE);
+
+  if (!dst.isValid()) {
+    return Status(1, "Destination tmp FS is not valid.");
+  }
+
+  auto blkCount =
+      ceil(static_cast<double>(src.size()) / static_cast<double>(kBuffSize));
+
+  for (size_t i = 0; i < blkCount; i++) {
+    std::vector<char> inBuff(kBuffSize, 0);
+    auto bytesRead = src.read(inBuff.data(), kBuffSize);
+    auto bytesWritten = dst.write(inBuff.data(), bytesRead);
+    if (bytesWritten < 0) {
+      return Status(1, "Error writing bytes to tmp fs");
+    }
+  }
+
+  return Status(0, "Ok");
+};
+
+Status Carver::compress(const std::set<boost::filesystem::path>& paths) {
+  auto arch = archive_write_new();
+  archive_write_set_format_zip(arch);
+  archive_write_set_format_pax_restricted(arch);
+  archive_write_open_filename(arch, archivePath_.string().c_str());
+
+  for (const auto& f : paths) {
+    PlatformFile pFile(f.string(), PF_OPEN_EXISTING | PF_READ);
+
+    auto entry = archive_entry_new();
+    archive_entry_set_pathname(entry, f.string().c_str());
+    archive_entry_set_size(entry, pFile.size());
+    archive_entry_set_filetype(entry, AE_IFREG);
+    archive_entry_set_perm(entry, 0644);
+    archive_write_header(arch, entry);
+
+    // TODO: Chunking or a max file size.
+    std::ifstream in(f.string(), std::ios::binary);
+    std::stringstream buffer;
+    buffer << in.rdbuf();
+    archive_write_data(arch, buffer.str().c_str(), buffer.str().size());
+    in.close();
+    archive_entry_free(entry);
+  }
+
+  archive_write_close(arch);
+  archive_write_free(arch);
```

Comment:
http://www.unix.com/man-page/freebsd/3/archive_write_free/

You probably don't need close as well if you're not checking the errors returned from it